File: internal/tools/registry.go
package tools

import "github.com/sagnikc395/kai/internal/api"

var registry = map[string]Tool{}

func init() {
	Register(BashTool{})
	Register(ReadTool{})
	Register(WriteTool{})
	Register(EditTool{})
	Register(GlobTool{})
	Register(GrepTool{})
}

func Register(tool Tool) {
	registry[tool.Name()] = tool
}

func Get(name string) (Tool, bool) {
	tool, ok := registry[name]
	return tool, ok
}

func All() []Tool {
	return []Tool{
		registry["bash"],
		registry["read"],
		registry["write"],
		registry["edit"],
		registry["glob"],
		registry["grep"],
	}
}

func Definitions() []api.ToolDefinition {
	definitions := make([]api.ToolDefinition, 0, len(All()))
	for _, tool := range All() {
		definitions = append(definitions, api.ToolDefinition{
			Type: "function",
			Function: api.ToolDefinitionFunction{
				Name:        tool.Name(),
				Description: tool.Description(),
				Parameters:  tool.Parameters(),
			},
		})
	}
	return definitions
}

File: kai/tools/registry.py
from __future__ import annotations

from typing import Any

from kai.tools.base import Tool
from kai.tools.bash import BashTool
from kai.tools.edit import EditTool
from kai.tools.glob_ import GlobTool
from kai.tools.grep_ import GrepTool
from kai.tools.read import ReadTool
from kai.tools.write import WriteTool

_registry: dict[str, Tool] = {}


def _init() -> None:
    _register(BashTool())
    _register(ReadTool())
    _register(WriteTool())
    _register(EditTool())
    _register(GlobTool())
    _register(GrepTool())


def _register(tool: Tool) -> None:
    _registry[tool.name()] = tool


def get(name: str) -> Tool | None:
    return _registry.get(name)


def all_tools() -> list[Tool]:
    return [
        _registry["bash"],
        _registry["read"],
        _registry["write"],
        _registry["edit"],
        _registry["glob"],
        _registry["grep"],
    ]


def definitions() -> list[dict[str, Any]]:
    defs: list[dict[str, Any]] = []
    for tool in all_tools():
        defs.append(
            {
                "type": "function",
                "function": {
                    "name": tool.name(),
                    "description": tool.description(),
                    "parameters": tool.parameters(),
                },
            }
        )
    return defs


_init()
